src/wp-includes/media.php

```diff
@@ -5651,104 +5624,154 @@ function wp_get_loading_optimization_attributes( $tag_name, $attr, $context ) {
+		 * to force-omit the attribute are other potential values).
 		 */
-		if ( doing_filter( 'the_content' ) ) {
-			return $loading_attrs;
+		if ( 'lazy' === $attr['loading'] ) {
+			$in_viewport = false;
+		} else {
+			$in_viewport = true;
 		}
+	}
 
-		// Conditionally skip lazy-loading on images before the loop.
-		if (
-			// Only apply for main query but before the loop.
-			$wp_query->before_loop && $wp_query->is_main_query()
+	// Logic to handle a `fetchpriority` attribute that is already provided.
+	if ( isset( $attr['fetchpriority'] ) && 'high' === $attr['fetchpriority'] ) {
+		/*
+		 * If the image was already determined to not be in the viewport (e.g.
+		 * from an already provided `loading` attribute), trigger a warning.
+		 * Otherwise, the value can be interpreted as in viewport, since only
+		 * the most important in-viewport image should have `fetchpriority` set
+		 * to "high".
+		 */
+		if ( false === $in_viewport ) {
+			_doing_it_wrong(
+				__FUNCTION__,
+				__( 'An image should not be lazy-loaded and marked as high priority at the same time.' ),
+				'6.3.0'
+			);
 			/*
-			 * Any image before the loop, but after the header has started should not be lazy-loaded,
-			 * except when the footer has already started which can happen when the current template
-			 * does not include any loop.
+			 * Set `fetchpriority` here for backward-compatibility as we should
+			 * not override what a developer decided, even though it seems
+			 * incorrect.
 			 */
-			&& did_action( 'get_header' ) && ! did_action( 'get_footer' )
-		) {
-			// Increase media count if there are images in header above a certian minimum size threshold.
-			$maybe_increase_content_media_count();
-			return $postprocess( $loading_attrs, true );
+			$loading_attrs['fetchpriority'] = 'high';
+		} else {
+			$in_viewport = true;
+		}
+	}
+
+	if ( null === $in_viewport ) {
+		$header_area = WP_TEMPLATE_PART_AREA_HEADER;
+
+		switch ( $context ) {
+			// Consider elements with these header-specific contexts to be in viewport.
+			case "template_part_{$header_area}":
+			case 'get_header_image_tag':
+				$in_viewport          = true;
+				$maybe_increase_count = true;
+				break;
+			// Count main content elements and detect whether in viewport.
+			case 'the_content':
+			case 'the_post_thumbnail':
+			case 'do_shortcode':
+				// Only elements within the main query loop have special handling.
+				if ( ! is_admin() && in_the_loop() && is_main_query() ) {
+					// Increase the counter since this is a main query content element.
+					$content_media_count = wp_increase_content_media_count();
+
+					// If the count so far is below the threshold, `loading` attribute is omitted.
+					if ( $content_media_count <= wp_omit_loading_attr_threshold() ) {
+						$in_viewport = true;
+					} else {
+						$in_viewport = false;
+					}
+				}
+				/*
+				 * For the 'the_post_thumbnail' context, the following case
```

Comment:
This inline doc is really helpful! If a shortcode is manually added prior to post content, shouldn't that be evaluated similarly to `the_post_thumbnail()`? Probably an edge case, but I could image a custom template trying to display a shortcode image above the content in a classic theme.

Reply:
We could consider that, but since it's a functional change, that should be in a separate ticket. It can be relevant, but agree it's probably an edge-case.